### ISSUE TITLE
Layout support

### ISSUE THREAD

**Issue by jureperak** [ltrzesniewski/RazorBlade#6]
Hi, i was playing around with this package and cannot figure out how would i make a .cshtml with Layout support? So i want to have two .cshtml and one of them would be a wrapper for the other.

e.g. in classic razor pages we are able to do in layout (lets call it SimpleLayout.cshtml)
```
<header>This is header</header>
@RenderBody()
<footer>This is footer</footer>
```

and then in view:
```
@{
    Layout = "SimpleLayout"
}

My content!
```

so the actual result would be:
```
This is header
My content!
This is footer
```

Greetings!

**Comment by ltrzesniewski** [comment]
Hello,

There is currently no layout support just like in ASP.NET, but that's a useful feature I suppose I should add eventually.

However, you can work around this limitation by writing two template files such as `Header.cshtml` and `Footer.cshtml`, then use them the following way:

```razor
@(new Header())
My content!
@(new Footer())
```

This doesn't handle sections, but similar workarounds should be possible for them by passing parameters to the `Header`/`Footer` templates.

**Comment by naasking** [comment]
This is a neat project. I've only just found this project so I'm not as familiar with all of the details, but from the type definitions I see, you could provide a efficient version of this for simple layout templates as follows:
```
@{
    Layout = new SimpleLayout();
}

My content!
```
The base definitions are changed slightly:
```
public interface IEncodedContent
{
    // the layout into which this content is rendered
    IEncodedContent? Layout { get; }
    // recursive, nested rendering pipeline
    void WriteTo(TextWriter textWriter, IEncodedContent? body);
}
```
The render procedure for a template then executes these "inside-out":
```
Layout?.Render(textWriter, this);
```
where the `body` parameter bound to `this` is available as a parameter in the razor file to execute in the appropriate place, eg. substitute `body.Render()` wherever `RenderBody()` is invoked. This nests arbitrarily because of the recursive design, so the layout template can itself have a layout template, ad infinitum.

This only supports a single template-body distinction, but it could possibly later be extended to provide full section support by changing the `body` parameter a `Dictionary<string, IEncodedContent>`, where the key is the section name.

As a brief aside, it would be helpful to have a more detailed breakdown or comparison of RazorBlade to [Razor slices](https://github.com/DamianEdwards/RazorSlices) which seems to have a similar goal, but slightly different design tradeoffs.

**Comment by ltrzesniewski** [comment]
Thanks! 🙂 

Yes, I considered implementing something like `Layout = new SimpleLayout();`, but not the way you describe, as `IEncodedContent` is somewhat similar to `IHtmlString` in ASP.NET, and I'd like to keep that interface simple.

I added a writer stack on another branch (07a829a9edde9737d486d7f1062d70a2f2c048f4) which could be useful for layouts. I'd also like to have full section support initially. I just need to find some time to work on this.

There are many existing Razor libraries, but I wasn't aware of RazorSlices. I wrote RazorBlade because I wanted two things I couldn't find elsewhere:
- a Razor library which does not reference anything from ASP.NET
- for the Razor templates to be compiled at build-time
